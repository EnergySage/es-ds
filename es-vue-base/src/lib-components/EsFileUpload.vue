<!-- eslint-disable vuejs-accessibility/no-static-element-interactions -->
<template>
    <div
        class="es-file-upload align-items-center rounded d-flex flex-column justify-content-center"
        :class="{ 'active': active, 'inactive': !active}"
        @dragenter.stop.prevent="active = true"
        @dragover.stop.prevent="active = true"
        @dragleave.stop.prevent="active = false"
        @dragend.stop.prevent="onDrop"
        @drop.stop.prevent="onDrop">
        <div
            class="d-flex justify-content-center"
            :class="{ 'flex-row': collapsed, 'flex-column': !collapsed }">
            <div class="align-self-center p-2">
                <icon-upload
                    class="d-none d-md-inline-block mb-4"
                    height="48px"
                    width="48px" />
            </div>
            <div class="align-self-center p-2">
                <slot name="cta" />
            </div>
            <div class="align-self-center p-2">
                <es-button
                    class="d-none d-md-inline-block w-35 mb-4"
                    variant="secondary"
                    outline
                    @click="openFilePicker">
                    Browse Files
                </es-button>
                <es-button
                    class="d-inline-block d-md-none w-100 mb-2"
                    variant="secondary"
                    outline
                    @click="openFilePicker">
                    Choose File
                </es-button>
            </div>
        </div>
        <div>
            <slot name="helpText" />
        </div>
        <b-form-file
            ref="fileInput"
            v-model="pickedItems"
            :state="Boolean(pickedItems)"
            :accept="fileTypesAsString"
            style="display: none"
            multiple />
    </div>
</template>

<script lang="js">
import { BFormFile } from 'bootstrap-vue';
import EsButton from './EsButton.vue';
import IconUpload from '../lib-icons/icon-upload.vue';

export default {
    name: 'EsFileUpload',
    components: {
        BFormFile,
        EsButton,
        IconUpload,
    },
    props: {
        uploadUrls: {
            type: Array,
            default: () => [],
        },
        fileTypes: {
            type: Array,
            default: () => [],
        },
        maxFileSize: {
            type: Number,
            default: 100,
            required: false,
        },
        uploadSuccessStatusCode: {
            type: Number,
            default: 200,
            required: false,
        },
        collapsed: {
            type: Boolean,
            default: false,
            required: false,
        },
    },
    data() {
        return {
            pickedItems: [],
            files: [],
            fileUrls: [], // For thumbnail preview
            active: false,
        };
    },
    computed: {
        fileTypesAsString() {
            return this.fileTypes.join(', ');
        },
    },
    watch: {
        uploadUrls: {
            handler(newUrls) {
                if (newUrls.length > 0 && newUrls.length === this.files.length) this.upload();
            },
            deep: true,
        },
        files: {
            handler(newFiles) {
                if (newFiles.length > 0 && newFiles.length === this.uploadUrls.length) this.upload();
            },
            deep: true,
        },
        pickedItems(newVal) {
            this.files = newVal;
            if (newVal.length > 0 && this.fileSizeValid()) {
                this.$emit('readyToUpload', this.files.length);
                this.readFilesIntoUrl(newVal);
            }
        },
    },
    methods: {
        fileSizeValid() {
            // Check if the combined file sizes exceed the maxFileSize prop
            const totalFileSize = this.files.reduce((acc, file) => acc + file.size, 0);
            if (totalFileSize > this.maxFileSize * 1000000) {
                this.$emit('fileSizeError');
                return false;
            }
            return true;
        },
        readFilesIntoUrl(files) {
            this.fileUrls = [];
            files
                .forEach((file) => {
                    const fileReader = new FileReader();
                    fileReader.onload = () => {
                        this.$emit('fileDataRead', {
                            name: file.name, type: file.type, size: file.size, data: fileReader.result,
                        });
                    };
                    fileReader.readAsDataURL(file);
                });
        },
        onDrop(event) {
            // The user has dropped files onto the component. We have to apply the same logic as if they had
            // selected the files from the file picker which limits the file types to the ones specified in the
            // fileTypes prop.
            this.active = false;
            this.files = [];

            // Use DataTransferItemList interface to access the file(s)
            const invalidFiles = [...event.dataTransfer.items].filter((item) => item.kind !== 'file');

            const dataTransfersAsFiles = [...event.dataTransfer.items]
                .filter((item) => item.kind === 'file')
                .map((item) => item.getAsFile());

            const validFiles = dataTransfersAsFiles
                .filter((file) => this.fileTypes.includes(file.type));
            const fileErrors = dataTransfersAsFiles
                .filter((file) => !this.fileTypes.includes(file.type))
                .concat(invalidFiles)
                .map((item) => item.name);
            if (fileErrors.length > 0) {
                this.$emit('fileTypeError', fileErrors);
            }

            this.files = validFiles;
            if (this.files.length > 0 && this.fileSizeValid()) {
                this.$emit('readyToUpload', this.files.length);
                this.readFilesIntoUrl(validFiles);
            }
        },
        openFilePicker() { this.$refs.fileInput.$el.childNodes[0].click(); },
        upload() {
            this.files.forEach((file, index) => {
                this.uploadSingleFile(file, index);
            });
        },
        uploadSingleFile(file, index) {
            const config = {
                headers: {
                    'Content-Type': file.type,
                },
                onUploadProgress: (progressEvent) => {
                    const percentCompleted = Math.round((progressEvent.loaded * 100) / progressEvent.total);
                    this.$emit('uploadProgress', { fileName: file.name, percentCompleted });
                    return percentCompleted;
                },
            };
            this.$axios.post(
                this.uploadUrls[index],
                file,
                config,
            )
                .then((response) => {
                    if (response.status === this.uploadSuccessStatusCode) {
                        this.$emit('uploadSuccess', file.name);
                    } else {
                        this.$emit('uploadFailure', file.name);
                    }
                })
                .catch(() => {
                    this.$emit('uploadFailure', file.name);
                });
        },
    },
};
</script>
<style lang="scss" scoped>
@import '~@energysage/es-bs-base/scss/bootstrap';
@import "~@energysage/es-bs-base/scss/variables";

.es-file-upload {
    background-color: $white;
    @include media-breakpoint-up(md) {
        background-color: $gray-200;
    }

    &.active {
        @media only screen and (min-width: 768px) {
            border: $btn-border-width dotted $cyan-500;
        }
    }

    &.inactive {
        @media only screen and (min-width: 768px) {
            border: $btn-border-width dotted $border-color;

        }
    }
}

</style>
