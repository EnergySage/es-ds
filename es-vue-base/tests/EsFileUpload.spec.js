import { mount } from '@vue/test-utils';
import EsFileUpload from '@/src/lib-components/EsFileUpload.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsFileUpload', () => {
    // Basic test; should exist for most components
    test('<EsFileUpload fileTypes={[\'image/png\']} />', async () => {
        const wrapper = mount(EsFileUpload, {
            ...jestVue,
            propsData: {
                fileTypes: ['image/png'],
            },
        });

        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.props('fileTypes')).toStrictEqual(['image/png']);
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    // Test slot is there
    test('<EsFileUpload fileTypes={[\'image/png\']}>My Content</EsFileUpload>', async () => {
        const wrapper = mount(EsFileUpload, {
            ...jestVue,
            slots: {
                default: 'My Content',
            },
            propsData: {
                fileTypes: ['image/png'],
            },
        });

        expect(wrapper.vm.$slots.default[0].text).toBe('My Content');
        expect(wrapper.html()).toMatchSnapshot();
    });
});
