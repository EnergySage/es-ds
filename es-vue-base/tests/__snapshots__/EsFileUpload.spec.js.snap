// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsFileUpload <EsFileUpload fileTypes={['image/png']} /> 1`] = `
"<div>
  <div role="presentation" class="es-file-upload align-items-center rounded d-flex flex-column justify-content-center p-2 mb-450">
    <div class="d-flex justify-content-center flex-column">
      <div class="align-self-center p-2"><svg viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg" style="height: 48px; width: 48px;" class="d-none d-md-inline-block mb-4">
          <path fill="currentColor" d="M1.5 30.734c.398 0 .78.182 1.06.506.282.324.44.763.44 1.22v8.634c0 .916.316 1.794.879 2.441.562.648 1.325 1.012 2.121 1.012h36c.796 0 1.559-.364 2.121-1.012.563-.647.879-1.525.879-2.441V32.46c0-.458.158-.897.44-1.221a1.408 1.408 0 0 1 1.06-.506c.398 0 .78.182 1.06.506.282.324.44.763.44 1.22v8.634c0 1.831-.632 3.588-1.757 4.883C45.117 47.272 43.59 48 42 48H6c-1.591 0-3.117-.728-4.243-2.023C.632 44.682 0 42.925 0 41.094V32.46c0-.458.158-.897.44-1.221a1.408 1.408 0 0 1 1.06-.506Z"></path>
          <path fill="currentColor" d="M22.94.507c.14-.16.305-.288.487-.375a1.331 1.331 0 0 1 1.15 0c.182.087.348.214.487.375l9 10.36c.282.324.44.764.44 1.222 0 .459-.158.898-.44 1.223a1.41 1.41 0 0 1-1.062.506 1.41 1.41 0 0 1-1.062-.506l-6.438-7.414V36.26c0 .458-.158.897-.44 1.221a1.408 1.408 0 0 1-1.06.506c-.398 0-.78-.182-1.06-.506a1.867 1.867 0 0 1-.44-1.22V5.897l-6.438 7.414c-.14.16-.305.287-.487.374a1.332 1.332 0 0 1-1.15 0 1.507 1.507 0 0 1-.487-.374c-.14-.161-.25-.352-.326-.561a1.958 1.958 0 0 1 0-1.323c.076-.21.186-.4.326-.561l9-10.36Z"></path>
        </svg></div>
      <div class="align-self-center p-2 text-center"></div>
      <div class="align-self-center p-2"><button type="button" class="btn w-100 w-md-auto mb-2 mb-md-4 btn-outline-secondary btn-md">
          <p class="d-none d-md-inline m-0">
            Browse files
          </p>
          <p class="d-md-none m-0">
            Choose file
          </p>
        </button></div>
    </div>
    <div></div>
    <div class="custom-file b-form-file is-valid" style="display: none;" id="__BVID__7__BV_file_outer_"><input type="file" accept="image/png" multiple="multiple" class="custom-file-input is-valid" style="z-index: -5;" id="__BVID__7"><label data-browse="Browse" class="custom-file-label" for="__BVID__7"><span class="d-block form-file-text" style="pointer-events: none;">No file chosen</span></label></div>
  </div>
  <!---->
  <div class="d-md-flex flex-nowrap mb-5"></div>
  <div id="preview-modal-___BV_modal_outer_" style="position: absolute; z-index: 1040;" class="modal-lightbox" body-class="border-top p-0">
    <transition-stub enterclass="" leaveclass="" entertoclass="" leavetoclass="" enteractiveclass="" leaveactiveclass="">
      <div id="preview-modal-" role="dialog" aria-hidden="true" aria-labelledby="preview-modal-___BV_modal_title_" aria-describedby="preview-modal-___BV_modal_body_" class="modal fade" style="display: none;">
        <div class="modal-dialog modal-xl modal-dialog-centered modal-dialog-scrollable">
          <!---->
          <div id="preview-modal-___BV_modal_content_" tabindex="-1" class="modal-content">
            <header id="preview-modal-___BV_modal_header_" class="modal-header">
              <h5 id="preview-modal-___BV_modal_title_" class="modal-title h2">

              </h5><button type="button" aria-label="Close" class="close"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" style="height: 26px; width: 26px;" class="text-gray-900">
                  <path fill-rule="evenodd" clip-rule="evenodd" d="M5.293 5.293a1 1 0 0 1 1.414 0L12 10.586l5.293-5.293a1 1 0 1 1 1.414 1.414L13.414 12l5.293 5.293a1 1 0 0 1-1.414 1.414L12 13.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L10.586 12 5.293 6.707a1 1 0 0 1 0-1.414Z"></path>
                </svg></button>
            </header>
            <div id="preview-modal-___BV_modal_body_" class="modal-body border-top p-0">
              <div class="d-md-none d-block file-preview text-center"><img lazy="true" src="" alt="" class="file-preview img-fluid"></div>
              <div class="d-none d-md-block">
                <div class="px-450 py-200 bg-gray-200">
                  <div class="card bg-white file-preview"><img lazy="true" src="" alt="" class="file-preview rounded-img img-fluid"></div>
                </div>
              </div>
            </div>
            <!---->
          </div>
          <!---->
        </div>
      </div>
    </transition-stub>
    <transition-stub name="" css="true" enterclass="" leaveclass="show" entertoclass="show" leavetoclass="" enteractiveclass="fade" leaveactiveclass="fade">
      <!---->
    </transition-stub>
  </div>
</div>"
`;

exports[`EsFileUpload <EsFileUpload fileTypes={['image/png']}>My Content</EsFileUpload> 1`] = `
"<div>
  <div role="presentation" class="es-file-upload align-items-center rounded d-flex flex-column justify-content-center p-2 mb-450">
    <div class="d-flex justify-content-center flex-column">
      <div class="align-self-center p-2"><svg viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg" style="height: 48px; width: 48px;" class="d-none d-md-inline-block mb-4">
          <path fill="currentColor" d="M1.5 30.734c.398 0 .78.182 1.06.506.282.324.44.763.44 1.22v8.634c0 .916.316 1.794.879 2.441.562.648 1.325 1.012 2.121 1.012h36c.796 0 1.559-.364 2.121-1.012.563-.647.879-1.525.879-2.441V32.46c0-.458.158-.897.44-1.221a1.408 1.408 0 0 1 1.06-.506c.398 0 .78.182 1.06.506.282.324.44.763.44 1.22v8.634c0 1.831-.632 3.588-1.757 4.883C45.117 47.272 43.59 48 42 48H6c-1.591 0-3.117-.728-4.243-2.023C.632 44.682 0 42.925 0 41.094V32.46c0-.458.158-.897.44-1.221a1.408 1.408 0 0 1 1.06-.506Z"></path>
          <path fill="currentColor" d="M22.94.507c.14-.16.305-.288.487-.375a1.331 1.331 0 0 1 1.15 0c.182.087.348.214.487.375l9 10.36c.282.324.44.764.44 1.222 0 .459-.158.898-.44 1.223a1.41 1.41 0 0 1-1.062.506 1.41 1.41 0 0 1-1.062-.506l-6.438-7.414V36.26c0 .458-.158.897-.44 1.221a1.408 1.408 0 0 1-1.06.506c-.398 0-.78-.182-1.06-.506a1.867 1.867 0 0 1-.44-1.22V5.897l-6.438 7.414c-.14.16-.305.287-.487.374a1.332 1.332 0 0 1-1.15 0 1.507 1.507 0 0 1-.487-.374c-.14-.161-.25-.352-.326-.561a1.958 1.958 0 0 1 0-1.323c.076-.21.186-.4.326-.561l9-10.36Z"></path>
        </svg></div>
      <div class="align-self-center p-2 text-center"></div>
      <div class="align-self-center p-2"><button type="button" class="btn w-100 w-md-auto mb-2 mb-md-4 btn-outline-secondary btn-md">
          <p class="d-none d-md-inline m-0">
            Browse files
          </p>
          <p class="d-md-none m-0">
            Choose file
          </p>
        </button></div>
    </div>
    <div></div>
    <div class="custom-file b-form-file is-valid" style="display: none;"><input type="file" accept="image/png" multiple="multiple" class="custom-file-input is-valid" style="z-index: -5;"><label data-browse="Browse" class="custom-file-label"><span class="d-block form-file-text" style="pointer-events: none;">No file chosen</span></label></div>
  </div>
  <!---->
  <div class="d-md-flex flex-nowrap mb-5"></div>
  <div id="preview-modal-___BV_modal_outer_" style="position: absolute; z-index: 1040;" class="modal-lightbox" body-class="border-top p-0">
    <transition-stub enterclass="" leaveclass="" entertoclass="" leavetoclass="" enteractiveclass="" leaveactiveclass="">
      <div id="preview-modal-" role="dialog" aria-hidden="true" aria-labelledby="preview-modal-___BV_modal_title_" aria-describedby="preview-modal-___BV_modal_body_" class="modal fade" style="display: none;">
        <div class="modal-dialog modal-xl modal-dialog-centered modal-dialog-scrollable">
          <!---->
          <div id="preview-modal-___BV_modal_content_" tabindex="-1" class="modal-content">
            <header id="preview-modal-___BV_modal_header_" class="modal-header">
              <h5 id="preview-modal-___BV_modal_title_" class="modal-title h2">

              </h5><button type="button" aria-label="Close" class="close"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="currentColor" style="height: 26px; width: 26px;" class="text-gray-900">
                  <path fill-rule="evenodd" clip-rule="evenodd" d="M5.293 5.293a1 1 0 0 1 1.414 0L12 10.586l5.293-5.293a1 1 0 1 1 1.414 1.414L13.414 12l5.293 5.293a1 1 0 0 1-1.414 1.414L12 13.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L10.586 12 5.293 6.707a1 1 0 0 1 0-1.414Z"></path>
                </svg></button>
            </header>
            <div id="preview-modal-___BV_modal_body_" class="modal-body border-top p-0">
              <div class="d-md-none d-block file-preview text-center"><img lazy="true" src="" alt="" class="file-preview img-fluid"></div>
              <div class="d-none d-md-block">
                <div class="px-450 py-200 bg-gray-200">
                  <div class="card bg-white file-preview"><img lazy="true" src="" alt="" class="file-preview rounded-img img-fluid"></div>
                </div>
              </div>
            </div>
            <!---->
          </div>
          <!---->
        </div>
      </div>
    </transition-stub>
    <transition-stub name="" css="true" enterclass="" leaveclass="show" entertoclass="show" leavetoclass="" enteractiveclass="fade" leaveactiveclass="fade">
      <!---->
    </transition-stub>
  </div>
</div>"
`;
